developer_tools/cargo/cargo.py

```diff
@@ -0,0 +1,562 @@
+    :rtype: bool
+    """
+    if unparsed_platform == "---":
+        return True
+
+    target_components = unparsed_platform.split("-")
+    return (
+        len(target_components) == 4
+        and rustc_cfg_dict["target_arch"] == target_components[0]
+        and rustc_cfg_dict["target_vendor"] == target_components[1]
+        and rustc_cfg_dict["target_os"] == target_components[2]
+        and rustc_cfg_dict["target_env"] == target_components[3]
+    )
+
+
+def parse_platform(unparsed_platform):
+    """
+    :param unparsed_platform: the string representation of the platform to parse
+    :type unparsed_platform: str
+    :returns: a bool indicating whether or not this platform should be included
+    based on the contents of rustc_cfg_dict
+    :rtype: bool
+    """
+
+    cfg_pattern = r"cfg\(([^-]*)\)"
+    cfg_re = re.compile(cfg_pattern)
+    cfg_match = cfg_re.match(unparsed_platform)
+
+    rustc_cfg_dict = build_rustc_cfg_dict()
+
+    if cfg_match is not None:
+        return parse_cfg_format_platform(cfg_match, rustc_cfg_dict)
+
+    return parse_target_format_platform(unparsed_platform, rustc_cfg_dict)
+
+
+def build_cargo_outdated_dict():
+    """
+    :returns: a dictionary containing information from the output of `cargo
+    outdated`
+    the keys are the string representations of dependencies
+    the values are 4-tuples containing
+    1) the string represenation of the dependency's version (i.e. from the
+    "Project" column of the output of `cargo outdated`,
+    2) the string representation of the dependency the dependency is pulled in by
+    or None if the dependency is not pulled in by any dependency
+    3) the string representation of the dependency's platform information (i.e.
+    from the "Platform" column of the ouput of `cargo outdated`
+    4) a bool indicating whether or not the dependency should be "included", with
+    respect to the platform information
+    :rtype: dict
+    """
+    cargo_outdated_dict = {}
+
+    command = ["cargo", "outdated"]
+    proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+
+    cargo_outdated_pattern = (
+        r"([^\s]*)\s*([^\s]*)\s*([^\s]*)\s*([^\s]*)\s*([^\s]*)\s*(.*)"
+    )
+    cargo_outdated_re = re.compile(cargo_outdated_pattern)
+
+    while True:
+        line = proc.stdout.readline()
+
+        if line == b"":
+            break
+
+        line_str = line.decode("utf-8")
+        cargo_outdated_match = cargo_outdated_re.match(line_str)
+
+        if cargo_outdated_match.group(1) in ("Name", "----"):
+            continue
+
+        dependencies = cargo_outdated_match.group(1)
+        dependencies_split = dependencies.split("->")
+
+        dependency = dependencies_split.pop(-1)
+
+        version = cargo_outdated_match.group(2)
+        pulled_in_by = None if dependencies_split == [] else dependencies_split[0]
+        platform = cargo_outdated_match.group(6)
+        include = parse_platform(platform)
+
+        cargo_outdated_dict[dependency] = (
```

Comment:
There is a difficulty here, because the ```dependency``` key is not necessarily unique. A dependency may be pulled in by multiple different packages and that dependency _may_ have different versions each time.

Reply:
This is true. The keys could be string-manipulated to include unique identifiers, perhaps. For example, "dependency-$-_i_", where _i_ is a natural number. @jbaublitz, what do you think of this idea?